index.js

```diff
@@ -60,48 +108,138 @@ app.post('/unregister', function(req, res) {
       res.sendStatus(404);
       return;
     }
-    client.del(token, function(err) {
-      res.sendStatus(200);
+    client.smembers(token, function(err, machines) {
+      for (var index = 0; index < machines.length; index++) {
+        client.del(machines[index]);
+      }
+      client.del(token, function(err) {
+        res.sendStatus(200);
+      });
+    });
+  });
+});
+
+// remove machine hash and its id from token set
+app.post('/unregisterMachine', function(req, res) {
+  var token = req.body.token;
+  var machineId = req.body.machineId;
+  console.log('Unregistering machine: ' + machineId);
+  client.exists(token, function(err, result) {
+    if (!result) {
+      res.sendStatus(404);
+      return;
+    }
+    client.exists(machineId, function(err, result) {
+      if (!result) {
+        res.sendStatus(404);
+        return;
+      }
+      client.del(machineId, function(err) {
+        client.srem(token, machineId, function(err) {
+          res.sendStatus(200);
+        });
+      });
     });
   });
 });
 
+
+// used only if registration is expired
+// it's needed as happens on request from service worker which doesn't
+// have any meta data
 app.post('/updateRegistration', function(req, res) {
   var token = req.body.token;
+  var machineId = req.body.machineId;
   client.exists(token, function(err, exists) {
     if (!exists) {
       res.sendStatus(404);
       return;
     }
-    client.hgetall(token, function(err, registration) {
-      client.hmset(token, {
-        "endpoint": req.body.endpoint,
-        "key": req.body.key
-      }, function() {
-        res.sendStatus(200);
+    client.sismember(token, machineId, function(err, ismember) {
+      if (!ismember) {
+        res.sendStatus(404);
+        return;
+      }
+      client.exists(machineId, function(err, exists) {
+        if (!exists) {
+          res.sendStatus(404);
+          return;
+        }
+        client.hmset(machineId, {
+          "endpoint": req.body.endpoint,
+          "key": req.body.key
+        }, function() {
+          res.sendStatus(200);
+        });
       });
     });
   });
 });
 
-app.post('/notify', function(req, res) {
+// used only if metadata updated
+// this happens on request to update meta data from front-end site
+app.post('/updateMeta', function(req, res) {
   var token = req.body.token;
+  var machineId = req.body.machineId;
   client.exists(token, function(err, exists) {
```

Comment:
This call should be unneeded, because the following `sismember` will fail if `token` doesn't exist.